### DISCUSSION TITLE
generalized csvwrite for subjects including subject name

### DISCUSSION THREAD

**Discussion by petersaj** [bonsai-rx/bonsai#1271] [Q&A]
Hi - I'm looking for generalized way to save multiple subjects into a single CSV file, ideally grabbing the name of each subject dynamically without having to externalize each one specifically.

For example: in a state machine workflow where each trial generates a variable number of outputs to subjects X and Y, I was thinking a CSV like this:

Trial 1, X, (X value 1), (timestamp)
Trial 1, X, (X value 2), (timestamp)
Trial 1, Y, (Y value 1), (timestamp)
Trial 2, X, (X value 1), (timestamp)
Trial 2, Y, (Y value 1), (timestamp)
Trial 2, Y, (Y value 2), (timestamp)

Is it possible to grab the subject name (e.g. X/Y) without specifically externalizing it? So the node connection could just look something like: 
![image](https://user-images.githubusercontent.com/17985056/222288096-88e6bf98-f362-4d32-81d8-2da3970b2e7e.png)

instead of something like: 
![image](https://user-images.githubusercontent.com/17985056/222288896-0ab1b2a1-c903-4e0b-8dbc-1f64c6d11160.png)

I thought BonVision's EventLogger and LogEvent might do something like this, but I don't think they use dynamic names either.

thanks,
Andy

**Comment by bruno-f-cruz** [comment]
Hi
Check #1226, I think it answers your question and offers a potential solution.
Cheers

**Reply by petersaj** [reply]
Thanks - I think a dictionary could do what I'm looking for, but I can't get the name of a subject to be used as a key without hardcoding it. In [#1226](https://github.com/orgs/bonsai-rx/discussions/1226), the keys are hardcoded in an **ExpressionTransform**: 

![image](https://user-images.githubusercontent.com/17985056/222528300-b1afaf96-cd73-488c-ab73-bb9ce20c3ff2.png)

or a **PythonTransform**:

![image](https://user-images.githubusercontent.com/17985056/222528413-a462b1a3-2352-4baa-acdc-69a88ed74b8e.png)

though I'm looking for something more like what the user originally was attempting, pulling out the name dynamically (which he said didn't work): 

![image](https://user-images.githubusercontent.com/17985056/222528738-b8fc0f89-c43b-4aee-8ba0-f5fdf2703dca.png)

I tried playing around with this kind of form: 
![image](https://user-images.githubusercontent.com/17985056/222529300-6ce9f638-9ecf-43f0-bc19-57d39f7eb28a.png)

where the **ExpressionTransforms** would output (Subject Name, Subject Value), but I couldn't find the commands to do that, is that something that might work?

**Reply by bruno-f-cruz** [reply]
I guess the point is that you cant create subjects "on-the-fly" without knowing their names during the building period. As an alternative, you can use observables and, using the same subject, filter its 
events with some sort of key.
Check two possible solutions attached, one using GroupBy and another one generating the expected keys in the stream.
[DynamicSubjects.zip](https://github.com/bonsai-rx/bonsai/files/10898217/DynamicSubjects.zip)

Your last example could be easily adapted to this motif.
